tests/quality/test_shrink_quality.py

```diff
@@ -204,19 +204,18 @@ def test_find_large_union_list():
     def large_mostly_non_overlapping(xs):
         assume(xs)
         assume(all(xs))
-        union = reduce(operator.or_, xs)
+        union = reduce(set.union, xs)
         return len(union) >= 30
 
     result = minimal(
         lists(sets(integers())),
         large_mostly_non_overlapping, timeout_after=120)
-    union = reduce(operator.or_, result)
+    union = reduce(set.union, result)
     assert len(union) == 30
     assert max(union) == min(union) + len(union) - 1
-    for x in result:
-        for y in result:
-            if x is not y:
-                assert not (x & y)
+    for i, x in enumerate(result[:-1]):
```

Comment:
Possibly a personal thing, but I find the old code easier to read, even if it’s less efficient.

You fixed the code after I wrote my original comment but before I hit post – which illustrates the value of simple, readable code! Original comment reproduced below for posterity.

---

You’ve also introduced an off-by-one error. This test fails for me locally with the following value for `result`:

```python
[{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
 {11, 12, 13, 14},
 {15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29}]
```

Your loop for `y` is emitting duplicate elements. You can see the difference if you run:

```python
result = [1, 2, 3]

for i, x in enumerate(result[:-1]):
    for y in result[i:]:
        print(i, x, y)
# (0, 1, 1)   !!!
# (0, 1, 2)
# (0, 1, 3)
# (1, 2, 2)   !!!
# (1, 2, 3)
```

vs

```python
result = [1, 2, 3]

for i, x in enumerate(result[:-1]):
    for y in result[i + 1:]:
        print(i, x, y)
# (0, 1, 2)
# (0, 1, 3)
# (1, 2, 3)
```

Rather than fixing the off-by-one error, if we’re not going for the slow-but-simple solution, I’d want to pass off the work to something like itertools. For example:

```python
import itertools

for x, y in itertools.combinations(result):
    assert not (x & y)
```

Reply:
`itertools.combinations` and `x.isdisjoint(y)` are much more elegant.